tests/test_all_reduce.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
```

Comment:
Is there enough overlap to test the collectives together in `test_collectives`? I imagine as they grow (and hopefully grow similar), the same utilities could be used to test both.

Reply:
There is, although I'm addressing this in a future PR as there are some nasty surprises. It might require further refinement of the python functions calling the collective ops.